src/shogun/clustering/Hierarchical.cpp

```diff
@@ -212,8 +216,94 @@ SGMatrix<int32_t> CHierarchical::get_cluster_pairs()
 	return SGMatrix<int32_t>(pairs,2,merges, false);
 }
 
-
 void CHierarchical::store_model_features()
 {
-	/* TODO. Currently does nothing since apply methods are not implemented. */
+	CDenseFeatures<float64_t>* rhs =
+	    dynamic_cast<CDenseFeatures<float64_t>*>(distance->get_rhs());
+
+	int32_t num_vectors = rhs->get_num_vectors();
+	int32_t num_features = rhs->get_num_features();
+
+	int32_t* n_cluster_samples = SG_MALLOC(int32_t, num_vectors + merges);
+	SGVector<int32_t>::fill_vector(n_cluster_samples, num_vectors + merges, 0);
+
+	SGMatrix<float64_t>* null_matrix =
+	    new SGMatrix<float64_t>(num_features, num_vectors + merges);
+	CDenseFeatures<float64_t>* temp_cluster_centers =
+	    new CDenseFeatures<float64_t>(*null_matrix);
+	delete null_matrix;
+
+	for (int32_t i = 0; i < num_vectors; i++)
+	{
+		bool dofree_c, dofree_rhs;
+		int32_t centerIdx = assignment[i];
+#ifdef DEBUG_HIERARCHICAL
+		SG_PRINT("\n");
+		SG_PRINT("On %04i point, with assignment=%04i \n", i, centerIdx);
+#endif
+		float64_t* center = temp_cluster_centers->get_feature_vector(
+		    centerIdx, num_features, dofree_c);
+		float64_t* sample_point =
+		    rhs->get_feature_vector(i, num_features, dofree_rhs);
+		for (int32_t j = 0; j < num_features; j++)
+		{
+			if (n_cluster_samples[centerIdx] > 0)
+			{
+				center[j] = (center[j] * n_cluster_samples[centerIdx] +
+				             sample_point[j]) /
+				            (n_cluster_samples[centerIdx] + 1);
+			}
+			else
+				center[j] = sample_point[j];
+#ifdef DEBUG_HIERARCHICAL
+			SG_PRINT(
```

Comment:
I think you could use `SG_DEBUG` instead? And then just set the global io to debug?

Reply:
Okay. Will make the change.